src/golang/cmd/server/handler/connect_integration.go

```diff
@@ -202,58 +207,241 @@ func (h *ConnectIntegrationHandler) Perform(ctx context.Context, interfaceArgs i
 			// Wait until the server is paused
 			h.PauseServer()
 			// Makes sure that the server is restarted
 			defer h.RestartServer()
 
-			// Wait until there are no more workflow runs in progress
-			lock := utils.NewExecutionLock()
-			if err := lock.Lock(); err != nil {
-				log.Errorf("Unexpected error when acquiring workflow execution lock: %v. Aborting storage migration!", err)
-				return
-			}
+			execState := storageMigrationObj.ExecState
+
+			var err error
 			defer func() {
-				if err := lock.Unlock(); err != nil {
-					log.Errorf("Unexpected error when unlocking workflow execution lock: %v", err)
+				if err != nil {
+					// Regardless of whether the migration succeeded, we should track the finished timestamp.
+					finishedAt := time.Now()
+					execState.Timestamps.FinishedAt = &finishedAt
+
+					execState.UpdateWithFailure(
+						// TODO: this can be a system error too. But no one cares right now.
+						shared.UserFatalFailure,
+						&shared.Error{
+							Tip:     fmt.Sprintf("Failure occurred when migrating to the new storage integration %s.", integrationObj.Name),
+							Context: err.Error(),
+						},
+					)
+					err = h.updateStorageMigrationExecState(ctx, storageMigrationObj.ID, &execState)
+					if err != nil {
+						log.Errorf("Unexpected error when updating the storage migration entry to FAILED: %v", err)
+						return
+					}
+
 				}
 			}()
 
-			if err := setIntegrationAsStorage(
-				context.Background(),
-				args.Service,
-				args.Config,
-				args.OrgID,
-				h.DAGRepo,
-				h.ArtifactRepo,
-				h.ArtifactResultRepo,
-				h.OperatorRepo,
-				h.IntegrationRepo,
-				h.Database,
-			); err != nil {
-				log.Errorf("Unexpected error when setting the new storage layer: %v", err)
+			// Mark the migration explicitly as RUNNING.
+			runningAt := time.Now()
+			execState.Timestamps.RunningAt = &runningAt
+			err = h.updateStorageMigrationExecState(ctx, storageMigrationObj.ID, &execState)
+			if err != nil {
+				log.Errorf("Unexpected error when updating the storage migration entry to RUNNING: %v", err)
+				return
+			}
+
+			// TODO: REMOVE
+			if strings.HasPrefix(integrationObj.Name, "failing") {
+				err = errors.Newf("I REALLY DONT LIKE YOUR NAME %s SIR", integrationObj.Name)
+				return
+			}
+
+			// Actually perform the storage migration.
+			storageConfig, err := h.performStorageMigration(args.Service, args.Config, args.OrgID)
+			// We let the defer() handle the failure case appropriately.
+			if err != nil {
+				return
 			}
 
 			log.Info("Successfully migrated the storage layer!")
+			finishedAt := time.Now()
+			execState.Timestamps.FinishedAt = &finishedAt
+			execState.Status = shared.SucceededExecutionStatus
+
+			// The update of the storage config and storage migration entry should happen together.
+			// While we don't enforce this atomically, we can make the two update together to minimize the risk.
+			err = h.updateStorageMigrationExecState(ctx, storageMigrationObj.ID, &execState)
```

Comment:
As we discussed a couple days ago, do we have a task that handles combining this logic into one, so the `config` pkg always reads from the DB to get the latest storage config.

Reply:
Yes will add.

Reply:
https://linear.app/aqueducthq/issue/ENG-2725/have-the-get-config-route-always-read-from-the-storage-migration-table